### PR TITLE
Bugfixes for HydrostaticSphericalCoriolis and solid body rotation validation experiment

### PR DESCRIPTION
**Pull request by glwagner**
This PR makes a few bugfixes and improvments:

1. Fixes a bug for `VectorInvariantEnstrophyConserving` Coriolis scheme for `HydrostaticSphericalCoriolis`
2. Edits tracer advection operators so they run on curvilinear grids
3. Adds `validation/solid_body_rotation/` based on [Williamson et al. 1992](https://www.sciencedirect.com/science/article/pii/S0021999105800166)

This PR also attempted to add a tracer advection test (using the solid body rotation solution), but this test fails. So we have some work to do to get tracer advection to work still.

It'd be nice to implement some quantitative measures of error for the solid body rotation test before merging.

Here's a movie that illustrates what the solid body rotation test does (showing incorrect tracer advection):

https://user-images.githubusercontent.com/15271942/109338288-0c9a3f80-7834-11eb-8189-9137cec2cd92.mp4

As a side note, it would be extremely useful to have a utility that computes a height field satisfying discrete geostrophic balance with some flow + Coriolis implementation.

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
A git merge/rebase master might help with review, but only thought I had was: would it be too much work to add tests to CI that run the two new validation experiment? This would help ensure they don't go out of date and that we can easily run them on Buildkite.

**Comment by glwagner** [comment]
> only thought I had was: would it be too much work to add tests to CI that run the two new validation experiment? This would help ensure they don't go out of date and that we can easily run them on Buildkite.

No, I think we should do this for the solid body validation experiment. I will implement this after I finish a quantitative analysis of the results.

File: src/Advection/centered_advective_fluxes.jl [hunk 1]
```diff
@@ -28,6 +28,6 @@ const Centered = AbstractCenteredAdvectionScheme
 ##### Advective tracer flux operators
 #####
     
-@inline advective_tracer_flux_x(i, j, k, grid, scheme::Centered, U, c) = @inbounds Axᵃᵃᶠ(i, j, k, grid) * U[i, j, k] * _symmetric_interpolate_xᶠᵃᵃ(i, j, k, grid, scheme, c)
-@inline advective_tracer_flux_y(i, j, k, grid, scheme::Centered, V, c) = @inbounds Ayᵃᵃᶠ(i, j, k, grid) * V[i, j, k] * _symmetric_interpolate_yᵃᶠᵃ(i, j, k, grid, scheme, c)
-@inline advective_tracer_flux_z(i, j, k, grid, scheme::Centered, W, c) = @inbounds Azᵃᵃᵃ(i, j, k, grid) * W[i, j, k] * _symmetric_interpolate_zᵃᵃᶠ(i, j, k, grid, scheme, c)
+@inline advective_tracer_flux_x(i, j, k, grid, scheme::Centered, U, c) = @inbounds Ax_uᶠᶜᶜ(i, j, k, grid, U) * _symmetric_interpolate_xᶠᵃᵃ(i, j, k, grid, scheme, c)
```
Comment (francispoulin) [inline src/Advection/centered_advective_fluxes.jl:31]: Introducing these new operators looks like a good idea.  One question is that in the old version, for `flux_x`, yu and `Ax^{aaf} * U`, whereas in the new version you have `Ax_u^{fcc}`, which is at a different location.  I presume this is one of the bugs you are fixing?

File: src/Advection/centered_second_order.jl [hunk 2]
```diff
@@ -22,6 +22,6 @@ const C2 = CenteredSecondOrder
 
 # Calculate the flux of a tracer quantity c through the faces of a cell.
 # In this case, the fluxes are given by u*Ax*T̅ˣ, v*Ay*T̅ʸ, and w*Az*T̅ᶻ.
-@inline advective_tracer_flux_x(i, j, k, grid, ::C2, U, c) = Ax_ψᵃᵃᶠ(i, j, k, grid, U) * ℑxᶠᵃᵃ(i, j, k, grid, c)
-@inline advective_tracer_flux_y(i, j, k, grid, ::C2, V, c) = Ay_ψᵃᵃᶠ(i, j, k, grid, V) * ℑyᵃᶠᵃ(i, j, k, grid, c)
-@inline advective_tracer_flux_z(i, j, k, grid, ::C2, W, c) = Az_ψᵃᵃᵃ(i, j, k, grid, W) * ℑzᵃᵃᶠ(i, j, k, grid, c)
+@inline advective_tracer_flux_x(i, j, k, grid, ::C2, U, c) = Ax_uᶠᶜᶜ(i, j, k, grid, U) * ℑxᶠᵃᵃ(i, j, k, grid, c)
```
Comment (francispoulin) [inline src/Advection/centered_second_order.jl:25]: Similarly these new functions seem to be at different positions and don't see a problem but just wanted to check if the old code was part of the bug

**Review by francispoulin — APPROVED** [review]
I have looked it over and think everything looks good, but admit that I haven't checked all the superscripts to make sure that everything is at the right location. I presume you have been careful about that and have checked everything, hence the bug fixes

**Comment by francispoulin** [comment]
The movie looks really intresting.  The first three spheres aren't doing much, which are `u,v,eta`, but `c` looks like solid body rotation.  But I thought you said that the tracer, `c`, was no working.  Should the first 3 be showing something more interesting?

**Comment by glwagner** [comment]
The dynamics test exhibits some "breathing":

https://user-images.githubusercontent.com/15271942/109686870-acb3d980-7b50-11eb-86c5-437fb7257d6b.mp4

https://user-images.githubusercontent.com/15271942/109686850-a887bc00-7b50-11eb-8939-803a1de29261.mp4

this could be because the initial condition doesn't satisfy discrete geostrophic balance, or a continuing issue with the coriolis implementation. However, I suggest we merge this PR now (since it has some bug fixes) and continue working on the validation tests in a future PR.

**Review by ali-ramadhan — APPROVED** [review]

**Pull request merged**